codegen/cmd/injection-gen/generators/reconciler_reconciler.go

```diff
@@ -0,0 +1,352 @@
+	// object. It is recommended that implementors do not call any update calls
+	// for the Kind inside of ReconcileKind, it is the responsibility of the calling
+	// controller to propagate those properties.
+	ReconcileKind(ctx context.Context, o *{{.type|raw}}) {{.reconcilerEvent|raw}}
+}
+
+// reconcilerImpl implements controller.Reconciler for {{.type|raw}} resources.
+type reconcilerImpl struct {
+	// Client is used to write back status updates.
+	Client {{.clientsetInterface|raw}}
+
+	// Listers index properties about resources
+	Lister {{.resourceLister|raw}}
+
+	// Recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	Recorder {{.recordEventRecorder|raw}}
+
+	// FinalizerName is the name of the finalizer to use when finalizing the
+	// resource.
+	FinalizerName string
+
+	// reconciler is the implementation of the business logic of the resource.
+	reconciler Interface
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*reconcilerImpl)(nil)
+
+`
+
+var reconcilerNewReconciler = `
+func NewReconciler(ctx context.Context, logger *{{.zapSugaredLogger|raw}}, client {{.clientsetInterface|raw}}, lister {{.resourceLister|raw}}, recorder {{.recordEventRecorder|raw}}, r Interface) {{.controllerReconciler|raw}} {
+	return &reconcilerImpl{
+		Client: client,
+		Lister: lister,
+		Recorder: recorder,
+		FinalizerName: defaultFinalizerName,
+		reconciler:    r,
+	}
+}
+`
+
+var reconcilerImplFactory = `
+// Reconcile implements controller.Reconciler
+func (r *reconcilerImpl) Reconcile(ctx context.Context, key string) error {
+	logger := {{.loggingFromContext|raw}}(ctx)
+
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := {{.cacheSplitMetaNamespaceKey|raw}}(key)
+	if err != nil {
+		logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+    // TODO: this is needed for serving.
+ 	// If our controller has configuration state, we'd "freeze" it and
+	// attach the frozen configuration to the context.
+	//    ctx = r.configStore.ToContext(ctx)
+
+	// Get the resource with this namespace/name.
+	original, err := r.Lister.{{.type|apiGroup}}(namespace).Get(name)
+	if {{.apierrsIsNotFound|raw}}(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logger.Errorf("resource %q no longer exists", key)
+		return nil
+	} else if err != nil {
+		return err
+	}
+	// Don't modify the informers copy.
+	resource := original.DeepCopy()
+
+	// Reconcile this copy of the resource and then write back any status
+	// updates regardless of whether the reconciliation errored out.
+	reconcileEvent := r.reconciler.ReconcileKind(ctx, resource)
+
+	// Synchronize the finalizers.
+	if equality.Semantic.DeepEqual(original.Finalizers, resource.Finalizers) {
+		// If we didn't change finalizers then don't call updateFinalizers.
+	} else if _, updated, fErr := r.updateFinalizers(ctx, resource); fErr != nil {
```

Comment:
This finalizer handling is racy.  I'd rather we rip out all of the finalizer bits here until we introduce an interface that lets us detect the intent to handle finalizers properly.

Reply:
it is now commented out and we will do a followup to add it back in. There are some ideas around reflecting the given reconciler to help understand what to do

Reply:
Removed.